### PR TITLE
Remove original ImageJ with add_legacy=False

### LINKED ISSUE #45: Disable original ImageJ support for the initial release
We should pass `add_legacy=False` to `imagej.init` for now, so that `no-legacy` commands appear in the search results. Otherwise, the functionality usable from napari-imagej is greatly reduced.

### PR DESCRIPTION
**Pull request by gselzer**
As noted by @ctrueden in #45, this will increase available functionality.

This PR is a draft because the removal of imagej-legacy also removes the dependency on `scijava-script`, which breaks tests.

Closes #45

### CODE REVIEW

**Comment by ctrueden** [comment]
What is `scijava-script`?

**Comment by gselzer** [comment]
> What is `scijava-script`?

Yeah, that was wrong. I am looking to keep the Python scripting capabilities in the dependencies; I am currently trying to figure out the needed dependencies. I think we need `scripting-python`, but it isn't a dependency of `imagej-legacy`...

**Comment by ctrueden** [comment]
You need `scripting-jython` I'm guessing, assuming you are running Jython scripts via `ij.script().run(...)` and similar.

**Comment by gselzer** [comment]
> You need `scripting-jython` I'm guessing, assuming you are running Jython scripts via `ij.script().run(...)` and similar.
> 
> The `scripting-python` component is something completely different, intended to leverage PyImageJ to execute CPython code from Java... not in production yet... will depend on the upcoming jep work.

Oh, gotcha.

My use case is wanting to write Python scripts that are wrapped into modules. Is that what scripting-jython does?

**Comment by ctrueden** [comment]
> My use case is wanting to write Python scripts that are wrapped into modules. Is that what scripting-jython does?

The SciJava script language plugins (`scijava/scripting-*` components) provide the ability to treat scripts in string form as SciJava modules, yeah. The `scripting-jython` component knows how to execute a Jython (not CPython!) script.

Where is the code you have that needs to do this? Or do I misunderstand what you are trying to do?

**Comment by gselzer** [comment]
> Where is the code you have that needs to do this?

There are two different use cases.

One is for convenience in the tests. See [here](https://github.com/imagej/napari-imagej/blob/0c955c858a85148a1815a1e6f96f44f5045c7b43/tests/test_module_utils.py#L757)

Another is locally, where I have ImageJ scripts that I use for holistic informal testing. I think this would be a nice thing to add support for in the future, much like Fiji's Script Editor.

**Comment by gselzer** [comment]
The current issue seems to be that, at the current state, `ScriptInfo.getLanguage()` returns null for the Python scripts, as [`ScriptLanguageIndex.getByExtension`](https://github.com/scijava/scijava-common/blob/309dc33241cefab2a2207ae31b32c8f137072996/src/main/java/org/scijava/script/ScriptInfo.java#L239) does not contain the `py` extension. I'm not sure why it is not in there...

**Comment by gselzer** [comment]
I think this is ready to go @ctrueden, but it is waiting on the release of scijava-common 2.88.2 due to your fix.

**Comment by gselzer** [comment]
So I updated this PR to be rebased over the latest master.

Some concerning points:
* By disabling imagej-legacy, we remove the `Source` and `Help` buttons. I thought that the `Source` button was especially nice. Should we consider moving the button functionality from imagej-legacy upstream @ctrueden? I don't see why the [`HelpSearchActionFactory`](https://github.com/imagej/imagej-legacy/blob/a03160552a945044a0c472ccde27d54bd76432af/src/main/java/net/imagej/legacy/search/HelpSearchActionFactory.java#L52) couldn't move upstream to *at least* imagej-common, if not scijava-common, and the [`SourceSearchActionFactory`](https://github.com/imagej/imagej-legacy/blob/master/src/main/java/net/imagej/legacy/search/SourceSearchActionFactory.java) couldn't (theoretically, as this one does depend on some `net.imagej.legacy` stuff) move to scijava-common.
* Everytime I search for functionality, I get these errors on printed to my console:
```
[ERROR] Cannot create plugin: org.scijava.plugins.scripting.jython.JythonScriptLanguage
[java.lang.Enum.toString] [ERROR] Cannot create plugin: org.scijava.plugins.scripting.jython.JythonScriptLanguage
[ERROR] Cannot create plugin: org.scijava.plugins.scripting.renjin.RenjinScriptLanguage
[java.lang.Enum.toString] [ERROR] Cannot create plugin: org.scijava.plugins.scripting.renjin.RenjinScriptLanguage
[ERROR] Cannot create plugin: org.scijava.plugins.scripting.jruby.JRubyScriptLanguage
[java.lang.Enum.toString] [ERROR] Cannot create plugin: org.scijava.plugins.scripting.jruby.JRubyScriptLanguage
[ERROR] Cannot create plugin: org.scijava.plugins.scripting.scala.ScalaScriptLanguage
[java.lang.Enum.toString] [ERROR] Cannot create plugin: org.scijava.plugins.scripting.scala.ScalaScriptLanguage
```
We should figure out why this happens. I feel like it had something to do with another release I was waiting on...imagej-common, maybe?

**Comment by gselzer** [comment]
>     * Everytime I search for functionality, I get these errors on printed to my console:


Adding the configuration option `-Dscijava.log.level=debug` yields errors like:

```
[DEBUG] Cannot create plugin: org.scijava.plugins.scripting.jython.JythonScriptLanguage
java.lang.IllegalArgumentException: No such script engine: jython
	at org.scijava.script.AdaptedScriptLanguage.findFactory(AdaptedScriptLanguage.java:149)
	at org.scijava.script.AdaptedScriptLanguage.<init>(AdaptedScriptLanguage.java:71)
	at org.scijava.plugins.scripting.jython.JythonScriptLanguage.<init>(JythonScriptLanguage.java:62)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at org.scijava.plugin.PluginInfo.createInstance(PluginInfo.java:304)
	at org.scijava.plugin.DefaultPluginService.createInstance(DefaultPluginService.java:234)
	at org.scijava.plugin.DefaultPluginService.createInstances(DefaultPluginService.java:223)
	at org.scijava.plugin.DefaultPluginService.createInstancesOfType(DefaultPluginService.java:214)
	at org.scijava.plugin.AbstractSingletonService.initInstances(AbstractSingletonService.java:135)
	at org.scijava.plugin.AbstractSingletonService.getInstances(AbstractSingletonService.java:74)
	at org.scijava.script.DefaultScriptService.initScriptLanguageIndex(DefaultScriptService.java:273)
	at org.scijava.script.DefaultScriptService.scriptLanguageIndex(DefaultScriptService.java:238)
	at org.scijava.script.DefaultScriptService.getIndex(DefaultScriptService.java:119)
	at org.scijava.script.ScriptService.canHandleFile(ScriptService.java:312)
	at org.scijava.script.ScriptFinder.createInfos(ScriptFinder.java:163)
	at org.scijava.script.ScriptFinder.scanDirectory(ScriptFinder.java:150)
	at org.scijava.script.ScriptFinder.findScripts(ScriptFinder.java:113)
	at org.scijava.script.DefaultScriptService.initScripts(DefaultScriptService.java:314)
	at org.scijava.script.DefaultScriptService.scripts(DefaultScriptService.java:256)
	at org.scijava.script.DefaultScriptService.lambda$initialize$0(DefaultScriptService.java:230)
	at org.scijava.object.ObjectIndex.resolvePending(ObjectIndex.java:385)
	at org.scijava.object.ObjectIndex.get(ObjectIndex.java:130)
	at org.scijava.object.ObjectIndex.getAll(ObjectIndex.java:119)
	at org.scijava.module.DefaultModuleService.getModules(DefaultModuleService.java:138)
	at org.scijava.search.module.ModuleSearcher.search(ModuleSearcher.java:79)
```

So then, we need to determine why this presents itself when we remove `imagej-legacy`...

File: src/napari_imagej/setup_imagej.py [hunk 1]
```diff
@@ -39,17 +39,17 @@ def imagej_init():
     )
     config.add_option(f"-Dimagej2.dir={os.getcwd()}")  # TEMP
     config.endpoints.append("io.scif:scifio:0.43.1")
+    config.endpoints.append("org.scijava:scijava-common:2.89.0")  # TEMP
```
Comment (gselzer) [inline src/napari_imagej/setup_imagej.py:42]: Is this necessary?

File: src/napari_imagej/setup_imagej.py [hunk 2]
```diff
@@ -39,17 +39,17 @@ def imagej_init():
...
     log_debug("Completed JVM Configuration")
 
     # Parse imagej settings
     ij_dir = setting("imagej_installation")
 
-    _ij = imagej.init(ij_dir_or_version_or_endpoint=ij_dir, mode=get_mode())
+    _ij = imagej.init(
+        ij_dir_or_version_or_endpoint=ij_dir, mode=get_mode(), add_legacy=False
```
Comment (gselzer) [inline src/napari_imagej/setup_imagej.py:49]: We could  just make this a setting. Do you think people would find that useful @ctrueden?

File: src/napari_imagej/widget_IJ2.py [hunk 3]
```diff
@@ -43,12 +44,77 @@ def __init__(self, viewer: Viewer):
             self.gui_button.clicked.connect(lambda: self.to_ij.setEnabled(True))
             self.gui_button.clicked.connect(lambda: self.from_ij.setEnabled(True))
 
+    @property
+    def gui(self) -> "jc.UserInterface":
+        ensure_jvm_started()
+        return ij().ui().getDefaultUI()
```
Comment (gselzer) [inline src/napari_imagej/widget_IJ2.py:50]: Add docstring

File: src/napari_imagej/widget_IJ2.py [hunk 4]
```diff
@@ -43,12 +44,77 @@ def __init__(self, viewer: Viewer):
...
+
+            @JOverride
+            def windowActivated(self, event):
+                pass
+
+            @JOverride
+            def windowDeactivated(self, event):
+                pass
+
+        window.addWindowListener(NapariAdapter())
```
Comment (gselzer) [inline src/napari_imagej/widget_IJ2.py:118]: Add docstring

File: tests/conftest.py [hunk 5]
```diff
@@ -15,10 +15,18 @@ def ij():
     return ij()
 
 
+@pytest.fixture()
+def viewer(make_napari_viewer) -> Generator[Viewer, None, None]:
+    yield make_napari_viewer()
```
Comment (gselzer) [inline tests/conftest.py:20]: Add docstring

File: tests/conftest.py [hunk 6]
```diff
@@ -15,10 +15,18 @@ def ij():
...
+
+
+# HACK: Something about the ImageJ2 GUI causes these unit tests to hang.
+# Let's force it to close, for now.
+
```
Comment (gselzer) [inline tests/conftest.py:25]: Remove this

File: tests/conftest.py [hunk 7]
```diff
@@ -15,10 +15,18 @@ def ij():
...
+
 @pytest.fixture
-def imagej_widget(make_napari_viewer) -> Generator[ImageJWidget, None, None]:
+def imagej_widget(viewer) -> Generator[ImageJWidget, None, None]:
     # Create widget
-    viewer: Viewer = make_napari_viewer()
     ij_widget: ImageJWidget = ImageJWidget(viewer)
 
     yield ij_widget
```
Comment (gselzer) [inline tests/conftest.py:32]: This could now be a one-liner

File: tests/conftest.py [hunk 8]
```diff
@@ -28,9 +36,13 @@ def imagej_widget(make_napari_viewer) -> Generator[ImageJWidget, None, None]:
 
 
 @pytest.fixture
-def gui_widget(make_napari_viewer) -> Generator[GUIWidget, None, None]:
+def gui_widget(viewer) -> Generator[GUIWidget, None, None]:
+    def mock_setting(value: str):
+        return {"imagej_installation": None, "choose_active_layer": True}[value]
+
```
Comment (gselzer) [inline tests/conftest.py:42]: Document this well

File: tests/conftest.py [hunk 9]
```diff
@@ -40,15 +52,15 @@ def gui_widget(make_napari_viewer) -> Generator[GUIWidget, None, None]:
...
     # monkeypatch settings
     def mock_setting(value: str):
         return {"imagej_installation": None, "choose_active_layer": False}[value]
 
     napari_imagej.widget_IJ2.setting = mock_setting
 
     # Create widget
-    viewer: Viewer = make_napari_viewer()
     widget: GUIWidget = GUIWidget(viewer)
```
Comment (gselzer) [inline tests/conftest.py:65]: Document this well

File: tests/test_IJ2GUIWidget.py [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
 
 
+@pytest.fixture()
+def asserter(qtbot) -> Callable[[Callable[[], bool]], None]:
+    def assertFunc(func: Callable[[], bool]):
+        # Let things run for up to a minute
+        qtbot.waitUntil(func, timeout=60000)
+
+    return assertFunc
+
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:23]: Document this

File: tests/test_IJ2GUIWidget.py [hunk 11]
```diff
@@ -58,6 +67,28 @@ def clean_layers_and_Displays(qtbot, ij):
...
+            self._passed = self.handler()
+            self._done_handling = True
+        except Exception:
+            self._passed = False
+
+    def is_done(self) -> bool:
+        return self._done_handling
+
+    def passed(self) -> bool:
+        return self._passed
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:89]: Document this.

Also, can we broaden its usage?

File: tests/test_IJ2GUIWidget.py [hunk 12]
```diff
@@ -182,112 +200,83 @@ def test_active_data_send(qtbot, ij, gui_widget: GUIWidget):
...
     assert not button.isEnabled()
 
     # Show the button
     qtbot.mouseClick(gui_widget.gui_button, Qt.LeftButton, delay=1)
-    qtbot.waitUntil(lambda: button.isEnabled())
+    asserter(lambda: button.isEnabled())
 
     # Add some data to ImageJ
-    sample_data = jc.ArrayImgs.bytes(10, 10, 10)
+    sample_data = jc.ArrayImgs.longs(10, 10, 10)
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:212]: Undo this change

File: tests/test_IJ2GUIWidget.py [hunk 13]
```diff
@@ -182,112 +200,83 @@ def test_active_data_send(qtbot, ij, gui_widget: GUIWidget):
...
         dialog = QApplication.activeWindow()
         buttons = dialog.findChildren(QPushButton)
         assert 2 == len(buttons)
         for button in buttons:
             if "ok" in button.text().lower():
-                qtbot.mouseClick(button, Qt.LeftButton, delay=1)
-                handler.done_handling()
+                button.clicked.emit()
                 return
         pytest.fail("Could not find the Ok button!")
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:238]: Can we just return `false` here? I don't know that this error is being propagated to the end.

File: tests/test_IJ2GUIWidget.py [hunk 14]
```diff
@@ -182,112 +200,83 @@ def test_active_data_send(qtbot, ij, gui_widget: GUIWidget):
...
     # Click the button
     qtbot.mouseClick(button, Qt.LeftButton, delay=1)
     # Assert that we are back to the original window i.e. that the popup was handled
-    qtbot.waitUntil(handler.has_been_handled)
-
-    # Assert that the data is now in Fiji
-    active_display = ij.display().getActiveDisplay()
-    assert isinstance(active_display, jc.ImageDisplay)
-    assert "test_to" == active_display.getName()
+    asserter(runnable.is_done)
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:246]: Could make this a one-liner...

File: tests/test_IJ2GUIWidget.py [hunk 15]
```diff
@@ -1,15 +1,27 @@
...
+from qtpy.QtCore import QRunnable, Qt, QThreadPool
 from qtpy.QtGui import QPixmap
 from qtpy.QtWidgets import QApplication, QDialog, QHBoxLayout, QMessageBox, QPushButton
 
 from napari_imagej.setup_imagej import JavaClasses, running_headless
 from napari_imagej.widget_IJ2 import FromIJButton, GUIButton, GUIWidget, ToIJButton
 
 
+@pytest.fixture()
+def asserter(qtbot) -> Callable[[Callable[[], bool]], None]:
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:17]: Move this into `conftest.py`

File: tests/test_module_utils.py [hunk 16]
```diff
@@ -693,8 +693,12 @@ def run_module_from_script(ij, tmp_path, script):
     return _module_utils._pure_module_outputs(module, unresolved_inputs)
 
 
+#     module = ij.py.run_script(language="Groovy", script=script, args=[])
+#     return _module_utils._module_outputs(module)
```
Comment (gselzer) [inline tests/test_module_utils.py:697]: Delete this

File: tests/test_module_utils.py [hunk 17]
```diff
@@ -693,8 +693,12 @@ def run_module_from_script(ij, tmp_path, script):
...
+
+
 script_zero_layer_zero_widget: str = """
-c = 1
+from net.imglib2.img.array import ArrayImgs
```
Comment (gselzer) [inline tests/test_module_utils.py:701]: What is the point of this change

**Review by gselzer — COMMENTED** [review]
Rough, but will be useful! Clean it up! :broom:

File: tests/conftest.py (hunk 6 above)
Comment (gselzer) [reply tests/conftest.py]: Removed in 11a5cb5

File: tests/test_module_utils.py (hunk 16 above)
Comment (gselzer) [reply tests/test_module_utils.py]: Removed in 11a5cb5

File: src/napari_imagej/setup_imagej.py (hunk 1 above)
Comment (gselzer) [reply src/napari_imagej/setup_imagej.py]: [Let's find out](https://github.com/imagej/napari-imagej/runs/7983832447?check_suite_focus=true)!

Comment (gselzer) [reply src/napari_imagej/setup_imagej.py]: So, turns out it is. Does this have something to do with pom-scijava @ctrueden?

Comment (ctrueden) [reply src/napari_imagej/setup_imagej.py:42]: It will only be necessary until we get `net.imagej:imagej:2.6.0` and `sc.fiji:fiji:2.6.0` released.

File: src/napari_imagej/setup_imagej.py (hunk 2 above)
Comment (gselzer) [reply src/napari_imagej/setup_imagej.py]: With more thinking, I'd prefer we do it later. We can then gauge how much of a priority it is and make sure it is done correctly.

But I did clean up the settings in 6d7c169

File: tests/conftest.py (hunk 7 above)
Comment (gselzer) [reply tests/conftest.py:29]: Eh, not really, since we still want to clean up the window.

File: tests/test_IJ2GUIWidget.py (hunk 12 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: Done!

File: src/napari_imagej/widget_IJ2.py (hunk 3 above)
Comment (gselzer) [reply src/napari_imagej/widget_IJ2.py]: Done

File: src/napari_imagej/widget_IJ2.py (hunk 4 above)
Comment (gselzer) [reply src/napari_imagej/widget_IJ2.py]: Also done

File: tests/conftest.py (hunk 5 above)
Comment (gselzer) [reply tests/conftest.py]: Done

File: tests/conftest.py (hunk 8 above)
Comment (gselzer) [reply tests/conftest.py]: Done

File: tests/conftest.py (hunk 9 above)
Comment (gselzer) [reply tests/conftest.py]: Done

File: tests/test_IJ2GUIWidget.py (hunk 10 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: Done

File: tests/test_IJ2GUIWidget.py (hunk 13 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: This code was removed, so no longer necessary

File: tests/test_IJ2GUIWidget.py (hunk 14 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: Made this a lot shorter

File: tests/test_IJ2GUIWidget.py (hunk 15 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: I did!

File: tests/test_module_utils.py (hunk 17 above)
Comment (gselzer) [reply tests/test_module_utils.py]: Now it never even happened! :magic_wand:

File: tests/test_IJ2GUIWidget.py (hunk 11 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: Resolved

File: src/napari_imagej/widget_IJ2.py [hunk 18]
```diff
@@ -43,12 +44,82 @@ def __init__(self, viewer: Viewer):
...
+            # First things first, show the GUI
+            ij().ui().showUI(self.gui)
+            # Then, add our custom settings to the User Interface
+            if ij().legacy and ij().legacy.isActive():
+                self._ij1_UI_setup()
+            else:
+                self._ij2_UI_setup()
+        # Later shows - the GUI is "visible", but the appFrame probably isn't
+        else:
+            self.gui.getApplicationFrame().setVisible(True)
```
Comment (gselzer) [inline src/napari_imagej/widget_IJ2.py:69]: We should update the documentation to reflect the fact that closing the GUI no longer prevents issues.

File: tests/conftest.py [hunk 19]
```diff
@@ -1,24 +1,32 @@
...
-import napari_imagej.widget_IJ2
+from napari_imagej import widget_IJ2
 from napari_imagej.widget import ImageJWidget
 from napari_imagej.widget_IJ2 import GUIWidget
 
 
 @pytest.fixture(scope="module")
 def ij():
+    """Fixture providing the ImageJ2 Gateway"""
     from napari_imagej.setup_imagej import ij
```
Comment (gselzer) [inline tests/conftest.py:15]: Move this import up

File: tests/test_IJ2GUIWidget.py [hunk 20]
```diff
@@ -28,7 +30,35 @@ def ImageDisplay(self):
...
+
+    widget_IJ2.request_values = newfunc
+
+    yield
+
+    widget_IJ2.request_values = oldfunc
+
+
+@pytest.fixture(autouse=True)
+def clean_layers_and_Displays(asserter, ij, viewer: Viewer):
```
Comment (gselzer) [inline tests/test_IJ2GUIWidget.py:61]: Document this

**Review by gselzer — COMMENTED** [review]
Nearly done!

File: src/napari_imagej/widget_IJ2.py (hunk 18 above)
Comment (gselzer) [reply src/napari_imagej/widget_IJ2.py:69]: Done

File: tests/conftest.py (hunk 19 above)
Comment (gselzer) [reply tests/conftest.py:15]: Actually, we can't do that as it would require the renaming of this function.

File: tests/test_IJ2GUIWidget.py (hunk 20 above)
Comment (gselzer) [reply tests/test_IJ2GUIWidget.py]: Done
